voltdbclient/api.go

```diff
@@ -0,0 +1,243 @@
+/* This file is part of VoltDB.
+ * Copyright (C) 2008-2016 VoltDB Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package voltdbclient
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+)
+
+// Conn is a single connection to a single node of a VoltDB database
+type Conn struct {
+	tcpConn  *net.TCPConn
+	connData *connectionData
+}
+
+// connectionData are the values returned by a successful login.
+type connectionData struct {
+	hostId      int32
+	connId      int64
+	leaderAddr  int32
+	buildString string
+}
+
+// NewConn creates an initialized, authenticated Conn.
+func NewConnection(user string, passwd string, hostAndPort string) (*Conn, error) {
+	var conn = new(Conn)
+	var err error
+	var raddr *net.TCPAddr
+	var login bytes.Buffer
+
+	if raddr, err = net.ResolveTCPAddr("tcp", hostAndPort); err != nil {
+		return nil, fmt.Errorf("Error resolving %v.", hostAndPort)
+	}
+	if conn.tcpConn, err = net.DialTCP("tcp", nil, raddr); err != nil {
+		return nil, err
+	}
+	if login, err = serializeLoginMessage(user, passwd); err != nil {
+		return nil, err
+	}
+	if err = conn.writeLoginMessage(login); err != nil {
+		return nil, err
+	}
+	if conn.connData, err = conn.readLoginResponse(); err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// Close a connection if open. A Conn, once closed, has no further use.
+// To open a new connection, use NewConnection.
+func (conn *Conn) Close() error {
+	var err error = nil
+	if conn.tcpConn != nil {
+		err = conn.tcpConn.Close()
+	}
+	conn.tcpConn = nil
+	conn.connData = nil
+	return err
+}
+
+// GoString provides a default printable format for Conn.
+func (conn *Conn) GoString() string {
+	if conn.connData != nil {
+		return fmt.Sprintf("hostId:%v, connId:%v, leaderAddr:%v buildString:%v",
+			conn.connData.hostId, conn.connData.connId,
+			conn.connData.leaderAddr, conn.connData.buildString)
+	}
+	return "uninitialized"
+}
+
+// Ping the database for liveness.
+func (conn *Conn) TestConnection() bool {
+	if conn.tcpConn == nil {
+		return false
+	}
+	rsp, err := conn.Call("@Ping")
+	if err != nil {
+		return false
+	}
+	return rsp.Status() == SUCCESS
+}
+
+// Call invokes the procedure 'procedure' with parameter values 'params'
+// and returns a pointer to the received Response.
+func (conn *Conn) Call(procedure string, params ...interface{}) (*Response, error) {
```

Comment:
if names are very similar to java method names that would be ideal as one can look at corresponding java side.